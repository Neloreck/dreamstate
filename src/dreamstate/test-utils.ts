/*
 * 'https://github.com/Neloreck/dreamstate'
 */

export * from "@/dreamstate/test-utils/utils/getReactConsumer";
export * from "@/dreamstate/test-utils/utils/getReactProvider";
export * from "@/dreamstate/test-utils/utils/nextAsyncQueue";

export * from "@/dreamstate/test-utils/registry/addServiceObserver";
export * from "@/dreamstate/test-utils/registry/getServiceObserversCount";
export * from "@/dreamstate/test-utils/registry/isServiceProvided";
export * from "@/dreamstate/test-utils/registry/registerService";
export * from "@/dreamstate/test-utils/registry/removeServiceObserver";
export * from "@/dreamstate/test-utils/registry/unRegisterService";
