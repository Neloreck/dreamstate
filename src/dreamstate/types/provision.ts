import type { ReactNode } from "react";

/**
 * Properties for the Dreamstate context manager provider.
 * Defines the props used for initializing and configuring the provider.
 */
export interface IProviderProps<T> {
  initialState?: T;
  partialHotReplacement?: boolean;
  children?: ReactNode;
}
