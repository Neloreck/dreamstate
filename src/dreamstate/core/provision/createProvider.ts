import { FunctionComponent } from "react";

import { createManagersObserver } from "@/dreamstate/core/observing/createManagersObserver";
import { TDreamstateService } from "@/dreamstate/types";

/**
 * Create component for manual provision without HOC/Decorator-like api.
 * Useful if your root is functional component or you are using createComponent api without JSX.
 */
export function createProvider(
  sources: Array<TDreamstateService>
): FunctionComponent<object> {
  return createManagersObserver(null, sources);
}
