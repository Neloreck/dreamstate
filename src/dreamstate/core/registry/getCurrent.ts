import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyObject, TDreamstateService } from "@/dreamstate/types";

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TDreamstateService<any>>(
  Service: T
): InstanceType<T> | null {
  return CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T> || null;
}
