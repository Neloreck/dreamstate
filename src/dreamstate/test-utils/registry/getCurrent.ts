import { CONTEXT_SERVICES_REGISTRY } from "@/dreamstate/core/internals";
import { TAnyContextManagerConstructor } from "@/dreamstate/types";

/**
 * Get current manager instance from registry.
 * Returns null if nothing is found.
 */
export function getCurrent<T extends TAnyContextManagerConstructor>(
  Service: T
): InstanceType<T> | null {
  return CONTEXT_SERVICES_REGISTRY.get(Service) as InstanceType<T> || null;
}
