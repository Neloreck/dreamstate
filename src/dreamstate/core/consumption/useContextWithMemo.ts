import { MutableRefObject, useCallback, useEffect, useRef, useState } from "react";

import { CONTEXT_STATES_REGISTRY, EMPTY_ARR } from "@/dreamstate/core/internals";
import { subscribeToManager } from "@/dreamstate/core/registry/subscribeToManager";
import { unsubscribeFromManager } from "@/dreamstate/core/registry/unSubscribeFromManager";
import { IContextManagerConstructor, TUpdateSubscriber } from "@/dreamstate/types";

/**
 * Use manager hook with subscribed updates.
 * Same like common useContext hook, but has memo checks.
 */
export function useContextWithMemo<T extends object, D extends IContextManagerConstructor<T>>(
  Manager: D,
  depsSelector: (context: T) => Array<any>
): T {
  const [ state, setState ] = useState(function() {
    return CONTEXT_STATES_REGISTRY.get(Manager) as T;
  });
  const observed: MutableRefObject<Array<any>> = useRef(depsSelector(state));

  const updateMemoState: TUpdateSubscriber<T> = useCallback(function(nextContext: T): void {
    // Calculate changes like react core does and fire change only after update.
    const nextObserved = depsSelector(nextContext);

    for (let it = 0; it < nextObserved.length; it ++) {
      if (observed.current[it] !== nextObserved[it]) {
        observed.current = nextObserved;
        setState(nextContext);

        return;
      }
    }
  }, EMPTY_ARR);

  useEffect(function() {
    subscribeToManager(Manager, updateMemoState);

    return function() {
      unsubscribeFromManager(Manager, updateMemoState);
    };
  });

  return state;
}
