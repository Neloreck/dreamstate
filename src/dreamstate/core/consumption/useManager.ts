import { useContext } from "react";

import { useContextWithMemo } from "@/dreamstate/core/consumption/useContextWithMemo";
import { IContextManagerConstructor, TAnyObject } from "@/dreamstate/types";

/**
 * Use manager hook, higher order wrapper for useContext.
 */
export function useManager<
  T extends TAnyObject,
  D extends IContextManagerConstructor<T, any>
>(
  Manager: D,
  depsSelector?: (context: D["prototype"]["context"]) => Array<any>
): D["prototype"]["context"] {
  if (depsSelector) {
    return useContextWithMemo(Manager, depsSelector);
  } else {
    return useContext(Manager.REACT_CONTEXT);
  }
}
