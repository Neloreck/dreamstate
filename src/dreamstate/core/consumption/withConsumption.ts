import { Consume } from "@/dreamstate/core/consumption/Consume";
import { IConsume } from "@/dreamstate/types";

/**
 * HOC alias for @Consume.
 */
export const withConsumption: IConsume = Consume as IConsume;
