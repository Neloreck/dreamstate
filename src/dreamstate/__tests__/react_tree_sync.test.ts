import { mount } from "enzyme";
import { createElement } from "react";
import { act } from "react-dom/test-utils";

import { withConsumption } from "@/dreamstate/core/consumption/withConsumption";
import { createProvider } from "@/dreamstate/core/provision/createProvider";
import { getCurrent } from "@/dreamstate/core/registry/getCurrent";
import { TestContextManager } from "@/fixtures";

describe("React tree for provided and consumed components.", () => {
  it("Should correctly update subscribed to functional provider elements view.", () => {
    const reactTree = mount(
      createElement(
        createProvider([ TestContextManager ]),
        {},
        createElement(withConsumption([ TestContextManager ])(function(props: object) {
          return JSON.stringify(props);
        } as any))
      )
    );

    const testContextManager: TestContextManager | null = getCurrent(TestContextManager);

    expect(reactTree).toMatchSnapshot("Initial state.");
    expect(testContextManager).not.toBeNull();

    act(() => {
      testContextManager!.setContext({ first: "updated", second: 101010, third: true });
    });

    expect(testContextManager!.context.first).toBe("updated");
    expect(testContextManager!.context.second).toBe(101010);
    expect(testContextManager!.context.third).toBe(true);

    reactTree.update();

    expect(reactTree).toMatchSnapshot("Updated state.");
  });
});
