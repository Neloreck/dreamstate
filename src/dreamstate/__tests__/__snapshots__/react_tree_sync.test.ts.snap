// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`React tree for provided and consumed components. Should correctly update subscribed to functional provider elements view.: Initial state. 1`] = `
<Memo(DS.Observer)>
  <DS.Consumer>
    <Component
      first="first"
      second={2}
      third={false}
    >
      {"first":"first","second":2,"third":false}
    </Component>
  </DS.Consumer>
</Memo(DS.Observer)>
`;

exports[`React tree for provided and consumed components. Should correctly update subscribed to functional provider elements view.: Updated state. 1`] = `
<Memo(DS.Observer)>
  <DS.Consumer>
    <Component
      first="updated"
      second={101010}
      third={true}
    >
      {"first":"updated","second":101010,"third":true}
    </Component>
  </DS.Consumer>
</Memo(DS.Observer)>
`;
